src/catchup/DownloadBucketsWork.cpp

```diff
@@ -45,14 +45,14 @@ DownloadBucketsWork::doWork()
         {
             return WORK_SUCCESS;
         }
-        else if (allChildrenWaiting())
-        {
-            return WORK_WAITING;
-        }
         else if (anyChildRaiseFailure())
         {
             return WORK_FAILURE_RETRY;
         }
+        else if (!anyChildRunning())
+        {
+            return WORK_WAITING;
```

Comment:
Note that WORK_WAITING state is also utilized here: this is useful in cases where some children have completed and some are waiting; parent will also go into waiting state, and will be scheduled when any of its children wakes it up.